charms/tensorboard-controller/tests/unit/test_operator.py

```diff
@@ -0,0 +1,131 @@
+        harness.add_relation_unit(rel_id, "app/0")
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_not_leader(
+        self, k8s_resource_handler: MagicMock, crd_resource_handler: MagicMock, harness: Harness
+    ):
+        """Test that charm waits if it's not the leader."""
+        harness.begin_with_initial_hooks()
+
+        assert isinstance(harness.charm.model.unit.status, WaitingStatus)
+        assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_no_gateway_info_relation(
+        self, k8s_resource_handler: MagicMock, crd_resource_handler: MagicMock, harness: Harness
+    ):
+        """Test that charm waits if the gateway info relation is missing."""
+        harness.set_leader(True)
+        harness.begin_with_initial_hooks()
+
+        assert isinstance(harness.charm.model.unit.status, WaitingStatus)
+        assert harness.charm.model.unit.status == WaitingStatus(
+            "Waiting for gateway info relation"
+        )
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_active(
+        self, k8s_resource_handler: MagicMock, crd_resource_handler: MagicMock, harness: Harness
+    ):
+        """Test that charm goes to active status if the gateway info relation exists."""
+        harness.set_leader(True)
+        self._setup_gateway_info_relation(harness)
+
+        harness.begin_with_initial_hooks()
+
+        assert isinstance(harness.charm.model.unit.status, ActiveStatus)
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_pebble_layer(
+        self,
+        k8s_resource_handler: MagicMock,
+        crd_resource_handler: MagicMock,
+        harness: Harness,
+    ):
+        """Test creation of Pebble layer."""
+        harness.set_leader(True)
+        model, name = "kubeflow", "kubeflow-gateway"
+        harness.set_model_name(model)
+        self._setup_gateway_info_relation(harness, model, name)
+        harness.begin_with_initial_hooks()
+        harness.container_pebble_ready("tensorboard-controller")
+        assert harness.charm.container.get_service("tensorboard-controller").is_running()
+        pebble_plan = harness.get_container_pebble_plan("tensorboard-controller")
+        assert pebble_plan
+        assert pebble_plan.services
+        pebble_plan_info = pebble_plan.to_dict()
+        assert pebble_plan_info["services"]["tensorboard-controller"]["command"] == "/manager"
+        test_env = pebble_plan_info["services"]["tensorboard-controller"]["environment"]
+        # there should be 2 environment variables
+        assert 2 == len(test_env)
```

Comment:
This is more stating a preference than saying something is right or wrong.  Use your judgement if you disagree.

This feels like a bit too much.  Checking that environment variables you expect imo makes sense (eg: `"ISTIO_GATEWAY" in test_env` or `test_env["ISTIO_GATEWAY"] == ___`), but asserting we have exactly 2 feels like overkill.  The reason being that if someone adds a third env var later, have they really invalidated this test?  I'm not sure.

I go back and forth on these, so just raising it in case you hadn't considered the thought